unity/unity_memory_info.c

```diff
@@ -0,0 +1,460 @@
+	mono_assembly_foreach((GFunc)AllocateMemoryForImageMemPool,  &iterationContext);
+
+	return NULL;
+}
+
+static void FreeMonoManagedHeap(MonoManagedHeap* heap)
+{
+	uint32_t i;
+
+	for (i = 0; i < heap->sectionCount; i++)
+	{
+		g_free(heap->sections[i].sectionBytes);
+	}
+
+	g_free(heap->sections);
+}
+
+typedef struct VerifyHeapSectionStillValidIterationContext
+{
+	MonoManagedMemorySection* currentSection;
+	gboolean wasValid;
+} VerifyHeapSectionStillValidIterationContext;
+
+static void VerifyHeapSectionIsStillValid(void* context, void* sectionStart, void* sectionEnd)
+{
+	VerifyHeapSectionStillValidIterationContext* iterationContext = (VerifyHeapSectionStillValidIterationContext*)context;
+	if (iterationContext->currentSection->sectionSize != (uint8_t*)(sectionEnd) - (uint8_t*)(sectionStart))
+		iterationContext->wasValid = FALSE;
+	else if (iterationContext->currentSection->sectionStartAddress != (uint64_t)(sectionStart))
+		iterationContext->wasValid = FALSE;
+
+	iterationContext->currentSection++;
+}
+
+static gboolean MonoManagedHeapStillValid(MonoManagedHeap* heap)
+{
+	MonoDomain* domain = mono_domain_get();
+	VerifyHeapSectionStillValidIterationContext iterationContext;
+	int currentSectionCount;
+
+	currentSectionCount = GC_get_heap_section_count();
+	currentSectionCount += MonoMemPoolNumChunks(domain->mp);
+	currentSectionCount += MonoImagesMemPoolNumChunks();
+
+	if (heap->sectionCount != currentSectionCount)
+		return FALSE;
+
+	iterationContext.currentSection = heap->sections;
+	iterationContext.wasValid = TRUE;
+
+	GC_foreach_heap_section(&iterationContext, VerifyHeapSectionIsStillValid);
+	
+	return iterationContext.wasValid;
+}
+
+// The difficulty in capturing the managed snapshot is that we need to do quite some work with the world stopped,
+// to make sure that our snapshot is "valid", and didn't change as we were copying it. However, stopping the world,
+// makes it so you cannot take any lock or allocations. We deal with it like this:
+//
+// 1) We take note of the amount of heap sections and their sizes, and we allocate memory to copy them into.
+// 2) We stop the world.
+// 3) We check if the amount of heapsections and their sizes didn't change in the mean time. If they did, try again.
+// 4) Now, with the world still stopped, we memcpy() the memory from the real heapsections, into the memory that we
+//    allocated for their copies.
+// 5) Start the world again.
+
+static void CaptureManagedHeap(MonoManagedHeap* heap)
+{
+	MonoDomain* domain = mono_domain_get();
+	SectionIterationContext iterationContext;
+
+	while(TRUE)
+	{
+		GC_call_with_alloc_lock(CaptureHeapInfo, heap);
+		GC_stop_world_external();
+
+		if (MonoManagedHeapStillValid(heap))
+			break;
+
+		GC_start_world_external();
+
+		FreeMonoManagedHeap(heap);
+	}
+	
+	iterationContext.currentSection = heap->sections;
+	GC_foreach_heap_section(&iterationContext, CopyHeapSection);
+
+	mono_domain_lock (domain);
```

Comment:
I guess we assume all the other stopped thread via `GC_stop_world_external` are not holding the domain lock? Should we just not take the domain lock here since all other threads are suspended? Or maybe take domain lock first?